packages/node/src/policy-gen/greedy-compartment.js

```diff
@@ -0,0 +1,161 @@
+  moduleDescriptor.source
+
+/**
+ * Returns a constructor for a subclass of `ses`' `Compartment` which injects
+ * package from policy overrides into compartment `imports` if not already
+ * present.
+ *
+ * This causes the new imports to be resolved and expanded as if their usage was
+ * detected. The aim is to cause a generated policy to reflect the state of the
+ * overrides, but not be _limited to_ those overrides (as would likely occur if
+ * we merged them verbatim). This hopefully prevents the end-user from needing
+ * to hand-craft a giant policy override.
+ *
+ * @param {CompartmentMapDescriptor} compartmentMap
+ * @param {LavaMoatPolicyOverrides} [policyOverride]
+ * @returns {typeof Compartment}
+ * @internal
+ */
+export const makeGreedyCompartment = (compartmentMap, policyOverride) => {
+  if (!policyOverride) {
+    return Compartment
+  }
+
+  return class GreedyCompartment extends Compartment {
+    /**
+     * Creates a "greedy" import hook if we can.
+     *
+     * @param {CompartmentOptions} [options]
+     */
+    constructor(options = {}) {
+      const { importHook, name } = options
+
+      if (!importHook || !name || name === ATTENUATORS_COMPARTMENT) {
+        super(options)
+        return
+      }
+
+      /** @type {ImportHook} */
+      const importHookWrapper = async (specifier) => {
+        const moduleDescriptor = await importHook(specifier)
+
+        if (name in compartmentMap.compartments) {
+          const compartmentDescriptor = compartmentMap.compartments[name]
+
+          for (const [
+            moduleDescriptorName,
+            { compartment: moduleDescriptorCompartmentName },
+          ] of entries(compartmentDescriptor.modules)) {
+            // ignore self
+            if (
+              moduleDescriptorCompartmentName &&
+              compartmentMap.compartments[moduleDescriptorCompartmentName] !==
+                compartmentDescriptor
+            ) {
+              const modCompartmentDescriptor =
+                compartmentMap.compartments[moduleDescriptorCompartmentName]
+              if (modCompartmentDescriptor) {
+                const canonicalName = getCanonicalName(
+                  modCompartmentDescriptor,
+                  compartmentMap.entry.compartment ===
+                    modCompartmentDescriptor.name
+                )
+                if (
+                  canonicalName &&
+                  canonicalName in (policyOverride?.resources ?? {}) &&
+                  compartmentDescriptor.compartments.has(
+                    moduleDescriptorCompartmentName
+                  )
+                ) {
+                  // this supports the "deprecated" RecordModuleDescriptor
+                  // format and the new SourceModuleDescriptor format. we
+                  // are futurists here at LavaMoat.
+                  //
+                  // note: `record` and `source` are non-extensible, which
+                  // is why the shallow clones are needed
```

Comment:
- [ ] `WeakMap` the `ModuleSource`

Reply:
@naugtur

I tried this, but then realized if we're keeping this cache for `ModuleSource`, we should also keep once for `ModuleSource.imports` since that prop must also become a new array.  then I realized we don't need the cache at all, and can instead _mutate_ objects instead of overwriting them _if_ they are extensible.  

initially, they will be frozen, and we must reassign those properties to extensible shallow clones.  if we encounter them again, they will be extensible, so we can mutate freely.

at any rate, after consulting with Kris, it does not seem like we should worry about it too much. we also do not need to re-freeze them.